tmt/package_managers/__init__.py

```diff
@@ -0,0 +1,179 @@
+import dataclasses
+import shlex
+import sys
+from collections.abc import Iterator
+from typing import TYPE_CHECKING, Callable, Optional, Union, overload
+
+import tmt
+import tmt.log
+import tmt.plugins
+import tmt.utils
+from tmt.utils import Command, CommandOutput, Path, ShellScript
+
+if TYPE_CHECKING:
+    from tmt.steps.provision import Guest
+
+    # Using TypeAlias and typing-extensions under the guard of TYPE_CHECKING,
+    # to avoid the necessity of requiring the package in runtime. This way,
+    # we can deal with it in build time and when running tests.
+    if sys.version_info >= (3, 10):
+        from typing import TypeAlias
+    else:
+        from typing_extensions import TypeAlias
+
+    #: A type of package manager names.
+    GuestPackageManager: TypeAlias = str
+
+
+class Package(str):
+    def to_script(self) -> ShellScript:
+        return ShellScript(shlex.quote(self))
+
+
+Installable = Union[Package, Path]
+
+
+PackageManagerClass = type['PackageManager']
+
+
+_PACKAGE_MANAGER_PLUGIN_REGISTRY: tmt.plugins.PluginRegistry[PackageManagerClass] = \
+    tmt.plugins.PluginRegistry()
+
+
+def provides_package_manager(
+        package_manager: str) -> Callable[[PackageManagerClass], PackageManagerClass]:
+    """
+    A decorator for registering package managers.
+
+    Decorate a package manager plugin class to register a package manager.
+    """
+
+    def _provides_package_manager(package_manager_cls: PackageManagerClass) -> PackageManagerClass:
+        _PACKAGE_MANAGER_PLUGIN_REGISTRY.register_plugin(
+            plugin_id=package_manager,
+            plugin=package_manager_cls,
+            logger=tmt.log.Logger.get_bootstrap_logger())
+
+        return package_manager_cls
+
+    return _provides_package_manager
+
+
+def find_package_manager(name: 'GuestPackageManager') -> 'PackageManagerClass':
+    """
+    Find a package manager by its name.
+
+    :raises GeneralError: when the plugin does not exist.
+    """
+
+    plugin = _PACKAGE_MANAGER_PLUGIN_REGISTRY.get_plugin(name)
+
+    if plugin is None:
+        raise tmt.utils.GeneralError(
+            f"Package manager '{name}' was not found in package manager registry.")
+
+    return plugin
+
+
+def escape_packages(*packages: Installable) -> Iterator[str]:
+    for package in packages:
+        if isinstance(package, Path):
+            yield str(package)
+
+        else:
+            yield shlex.quote(package)
+
+
+# TODO: find a better name, "options" is soooo overloaded...
+@dataclasses.dataclass(frozen=True)
+class Options:
```

Comment:
`PackageManagerOptions`? The name is not linked to a cli variable name or anything right?

Reply:
Correct, it's not "options" as in "CLI options". This is just a collection of various controls, more or less resembling controls one can pass to a package manager. It's related to a single action.

Reply:
Oh, but the field names inside this  would be accessible from `.fmf`, e.g.:
```yaml
prepare:
  - how: install
    package: my-package
    install_root: /path/to/other/install/root
```

Would be worth at least adding a docstring for the class to clarify that.

Reply:
I never expected this class to be exposed to a user directly. In my view, it's rather an implementation detail, something a plugin like `install` would construct when calling a package manager. If the plugin needs to call `package_manager.install()` two times, both calls may get a very different set of options, depending on what's the higher goal of the plugin. The first call might fail, for example, the second might be more tolerant, exclude some packages, and so on. Basically, I wanted to avoid having this set of names as parameters of `install()`, `reinstall()` and other package manager methods - I'm pretty sure this is not the final set of controls, adding new ones would require changing dozens of methods. So, hence this internal bundle.

I'll update the docstring accordingly.

Reply:
Hmm, that sounds more like a `PackageManagerMeta`? I.e. where you define what and how the `PackageManager` class itself is capable of doing?

Reply:
Hmm, I wouldn't say so. They do describe additional parameters of one particular action, not a set of capabilities of a package manager. Something like "install packages `foo` and `bar`, but `check_first` and `exclude` package `bar`". They are indeed `install()` or `reinstall()` parameters, but bundled into a class we can extend without messing with said methods' signatures.

Reply:
So far we have `Instructions`, which I'm not very fond of as it contains mostly additional controls rather than actual instruction to install something :/ Keeping `Options` for now, it's easy to refactor.

Reply:
I suppose we may end up with `uninstall`, so I'd prefer to avoid `Install`. Anyway, we can let it be, and get back to it with smart ideas later, renaming this structure will be trivial.

Reply:
Ack, let's keep `Options`.